keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -0,0 +1,760 @@
+
+Credentials may also be specified at more granular levels in order to allow for context dependent keys.  E.g. When authorization differs between BucketClasses or between individual operations.  This may be facilitated by defining a set of string keys which the core automation will be aware of, so that Secrets may be referenced by BucketClasses.  For example:
+
+```
+cosi.io/provisioner-secret-name:
+cosi.io/provisioner-secret-namespace:
+```
+
+- **Per BucketClass:** A secret may be made specific to a BucketClass.  This suits cases where authorization may be segregated in the object store.  The Secret may then be defined explicitly in the `bucketClass.parameters` map.
+
+  ```
+  cosi.io/provisioner-secret-name: "foo"
+  cosi.io/provisioner-secret-namespace: "bar"
+  ```
+
+- **Per Operation/Bucket:** Unique credentials are passed per Bucket or operation. In order to support dynamic Secret naming, templating similar to [CSI Secret templating](https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html) may be used.  E.g.
+
+  ```
+  "${bucket.name}"
+  "${bucket.namespace}"
+  ```
+  
+  Admins may then define a BucketClass with the following parameters included:
+
+  *Per Bucket Operation*
+  
+  ```
+  cosi.io/provisioner-secret-name: "${bucket.name}"
+  cosi.io/provisioner-secret-namespace: "${bucket.namespace}"
+  ```
+---
+
+# gRPC Definitions
+
+There is one service defined by COSI - `Provisioner`. This will need to be satisfied by the vendor-provisioner in order to be COSI-compatible
+
+```
+service Provisioner {
+    rpc ProvisionerGetInfo (ProvisionerGetInfoRequest) returns (ProvisionerGetInfoResponse) {}
+
+    rpc ProvisionerCreateBucket (ProvisionerCreateBucketRequest) returns (ProvisionerCreateBucketResponse) {}
+    rpc ProvisionerDeleteBucket (ProvisionerDeleteBucketRequest) returns (ProvisionerDeleteBucketResponse) {}
+
+    rpc ProvisionerGrantBucketAccess (ProvisionerGrantBucketAccessRequest) returns (ProvisionerGrantBucketAccessResponse);
+    rpc ProvisionerRevokeBucketAccess (ProvisionerRevokeBucketAccessRequest) returns (ProvisionerRevokeBucketAccessResponse);
+}
+```
+
+## ProvisionerGetInfo
+
+This call is meant to retrieve the unique provisioner Identity. This identity will have to be set in `BucketRequest.Provisioner` field in order to invoke this specific provisioner.
+
+```
+message ProvisionerGetInfoRequest {
+    // Intentionally left blank
+}
+
+message ProvisionerGetInfoResponse {    
+    string provisioner_identity = 1;
+}
+```
+
+## ProvisonerCreateBucket
+
+This call is made to create the bucket in the backend. If the bucket already exists, then the appropriate error code `ALREADY_EXISTS` must be returned by the provisioner.
```

Comment:
If you call CreateBucket for the 1st time to create a new bucket, is it a blocking call until the bucket is created?  What does it return at the end?

Reply:
Here is S3 `CreateBucket` call - https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateBucket.html

The return value is essentially only `200 OK`

Reply:
`ALREADY_EXISTS` in CSI spec means a resource with the given name exists but incompatible with other input parameters.  We should do the same thing here.  So if CreateBucket creates a new bucket and returns OK when it is done, calling CreateBucket again should also return OK if bucket exists and compatible with all input parameters. This way CreateBucket also works like query bucket.  If for some reason the first call to CreateBucket times out, the controller can make another call to CreateBucket to check if it is created.

https://github.com/container-storage-interface/spec/blob/master/spec.md
```
Condition: Volume already exists but is incompatible
gRPC code: 6 ALREADY_EXISTS
Description: Indicates that a volume corresponding to the specified volume name already exists but is incompatible with the specified capacity_range, volume_capabilities, parameters, accessibility_requirements or volume_content_source.
```

Reply:
Updated KEP based on recommendation